middleware/rate_limiter.go

```diff
@@ -157,53 +157,39 @@ type (
 )
 
 // NewRateLimiterMemoryStore returns an instance of RateLimiterMemoryStore
-func NewRateLimiterMemoryStore(
-	options ...RateLimiterMemoryStoreOption,
-) (store *RateLimiterMemoryStore) {
-	store = &RateLimiterMemoryStore{
-		visitors:  make(map[string]*Visitor),
-		expiresIn: 3 * time.Minute,
-	}
-	for _, option := range options {
-		option(store)
-	}
-	if store.expiresIn == 0 {
-		store.expiresIn = 3 * time.Minute
+func NewRateLimiterMemoryStore(config RateLimiterMemoryStoreConfig) (store *RateLimiterMemoryStore) {
+	store = &RateLimiterMemoryStore{}
+
+	store.rate = config.rate
+	store.burst = config.burst
+	if config.expiresIn == 0 {
+		store.expiresIn = DefaultRateLimiterMemoryStoreConfig.expiresIn
+	} else {
+		store.expiresIn = config.expiresIn
 	}
+	store.visitors = make(map[string]*Visitor)
 	store.lastCleanup = now()
 	return
 }
 
-// RateLimiterMemoryStoreOption represents an option of RateLimiterMemoryStore
-type RateLimiterMemoryStoreOption func(store *RateLimiterMemoryStore)
-
-// RateLimiterMemoryStoreRate sets rate of RateLimiterMemoryStore
-func RateLimiterMemoryStoreRate(rate rate.Limit) RateLimiterMemoryStoreOption {
-	return func(store *RateLimiterMemoryStore) {
-		store.rate = rate
-	}
+// RateLimiterMemoryStoreConfig represents configuration for RateLimiterMemoryStore
+type RateLimiterMemoryStoreConfig struct {
+	rate      rate.Limit
+	burst     int
+	expiresIn time.Duration
 }
 
-// RateLimiterMemoryStoreBurst sets rate of RateLimiterMemoryStore
-func RateLimiterMemoryStoreBurst(burst int) RateLimiterMemoryStoreOption {
-	return func(store *RateLimiterMemoryStore) {
-		store.burst = burst
-	}
-}
-
-// RateLimiterMemoryStoreExpiresIn sets rate of RateLimiterMemoryStore
-func RateLimiterMemoryStoreExpiresIn(expiresIn time.Duration) RateLimiterMemoryStoreOption {
-	return func(store *RateLimiterMemoryStore) {
-		store.expiresIn = expiresIn
-	}
+// DefaultRateLimiterMemoryStoreConfig provides default configuration values for RateLimiterMemoryStore
+var DefaultRateLimiterMemoryStoreConfig = RateLimiterMemoryStoreConfig{
+	expiresIn: 3 * time.Minute,
 }
 
 // Allow implements RateLimiterStore.Allow
 func (store *RateLimiterMemoryStore) Allow(identifier string) bool {
-	store.mutex.Lock()
 	if now().Sub(store.lastCleanup) > store.expiresIn {
```

Comment:
The access to store.lastCleanup is racy again here with the access in cleanupStaleVisitors() to `store.lastCleanup = now()`

Reply:
I moved the lock to the cleanup goroutine

Reply:
That is true, the cleanup itself is locked, so cleanup will run async and ensure correct concurrent access.. but the requests coming in will check store.lastCleanup without a mutex, so multiple requests will read the old value and start an async cleanup and this will lead to issues.

Reply:
hmm yeah... how do you propose we resolve this?

use a second mutex? because double locking a mutex could cause a deadlock..

or should I make it synchronous again... and we explore the sharding process you talked about?

Reply:
but if I lock twice and unlock twice.. that could actually resolve it.

Reply:
To be on the safe side reusing our mutex we would need to do the mutex earlier.
The lock in the cleanupRouting will have to wait until the current request is finished (could be multiple if their locks are faster). After blocking in the async cleanup the lock will engage and cleanup will run and effectively block all other requests until it is finished (which makes it nearly sync anyway)

```go
func (store *RateLimiterMemoryStore) Allow(identifier string) bool {
	store.mutex.Lock()
	if now().Sub(store.lastCleanup) > store.expiresIn {
		go store.cleanupStaleVisitors() // to Lock() in will wait for the current lock to clear
	}
```go
	
But as you said, improving the store cleanup can be done in another PR